### PR TITLE
Optimize `brew update` by batching `git cat-file` operations

### LINKED ISSUE #13224: `brew update` can be terribly slow for stale taps
### `brew config` output

```shell
HOMEBREW_VERSION: 3.4.10-3-gbb3d970-dirty
ORIGIN: https://github.com/Homebrew/brew
HEAD: bb3d97030c68ed6d3074b4689a39a0ebb556fbf4
Last commit: 6 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: dbc32b7ea38309f8b0b91e9f1ecf68e17c8a519d
Core tap last commit: 8 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_MAKE_JOBS: 16
HOMEBREW_NO_ANALYTICS: set
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 13.1.6 build 1316
Git: 2.35.1 => /usr/local/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.3.1-x86_64
CLT: 13.3.0.0.1.1645755326
Xcode: N/A
```

### `brew doctor` output

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  bazaar

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  kubernetes-cli

Warning: You have uncommitted modifications to Homebrew/brew.
If this is a surprise to you, then you should stash these modifications.
Stashing returns Homebrew to a pristine state but can be undone
should you later need to do so for some reason.
  cd /usr/local/Homebrew && git stash && git clean -d -f

Uncommitted files:
   M Library/Homebrew/cmd/update-report.rb
   M Library/Homebrew/cmd/update.sh
```

### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

If you run `brew update` on a very stale tap, the `update-report` takes a long time.

For example,
```sh
git -C "$(brew --repo homebrew/core)" reset --hard e27948a0f03 # commit on master from Dec 2021
time brew update
```
takes **2m40s** on my MacBook Pro (2.3 GHz 8-Core Intel Core i9)

If I comment out _just_ [the call to `brew update-report`](https://github.com/Homebrew/brew/blob/13735bc7807d134df105de82acf82509c88227f9/Library/Homebrew/cmd/update.sh#L760), the same operation takes **3.8s** — so the expense is entirely in `brew update-report`.

Alternately, if I comment out _just_ [lines 373-383 in `update-report.rb`](https://github.com/Homebrew/brew/blob/f811774cab139ae83bccd379b4352a44eb22ead5/Library/Homebrew/cmd/update-report.rb#L373-L383), the same operation takes **10.3s**.

`brew update --preinstall` also [bypasses these lines](https://github.com/Homebrew/brew/blob/f811774cab139ae83bccd379b4352a44eb22ead5/Library/Homebrew/cmd/update-report.rb#L367-L371); but it cuts corners in `update.sh` as well.

### What happened (include all command output)?

(Described above)

### What did you expect to happen?

I would expect `brew update` not to take minutes, even for stale taps

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
(Described above)
```

### PR DESCRIPTION
**Pull request by boblail**
This approach is almost as optimal as [omitting this filtering altogether](https://github.com/Homebrew/brew/pull/13243).

It relies on `git cat-file --batch`, which can receive a list of blobs to look up on standard input.

| Homebrew Branch | Command | Commits Rewound | Time Elapsed | Updated Formulae |
| --- | --- | --- | --- | --- |
| `master` | `git -C "$(brew --repo homebrew/core)" reset --hard aa1b3d5df78; time sh -c "brew update &> output-2k.txt"` | 2,000 | 35.6s | 666 |
| `master` | `git -C "$(brew --repo homebrew/core)" reset --hard 11e6919661c; time sh -c "brew update &> output-20k.txt"` | 20,000 | 129.8s | 2411 |
| #13234 | `git -C "$(brew --repo homebrew/core)" reset --hard aa1b3d5df78; time sh -c "brew update &> output-2k-unfiltered.txt"` | 2,000 | 6.6s | 819 |
| #13234 | `git -C "$(brew --repo homebrew/core)" reset --hard 11e6919661c; time sh -c "brew update &> output-20k-unfiltered.txt"` | 20,000 | 10.4s | 3532 |
| _this one_ | `git -C "$(brew --repo homebrew/core)" reset --hard aa1b3d5df78; time sh -c "brew update &> output-2k-optimized.txt"` | 2,000 | 7.4s | 666 |
| _this one_ | `git -C "$(brew --repo homebrew/core)" reset --hard 11e6919661c; time sh -c "brew update &> output-20k-optimized.txt"` | 20,000 | 17.6s | 2411 |

Fixes #13224

### Alternatives

- https://github.com/Homebrew/brew/pull/13243

-----

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

**Comment by boblail** [comment]
@MikeMcQuaid, this approach makes lines like this in the specs inert:
```ruby
allow(FormulaVersions).to receive(:new).and_return(double(formula_at_revision: "2.0"))
```

If you prefer this approach to #13243, what'd be your inclination for architecture?

My first thought was to extract a class like this — that could be stubbed in a similar way:
```ruby
class FormulaeAtRevision
  def initialize(formulae, revision)
    # ...
  end

  # accepts the current version of a formula, return its version at `@revision`
  def version_of(formula)
    # ... 
  end
end
```

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]
Appreciate the work here but would love to avoid having a new class just for this lookback and instead look at improving `FormulaVersions` to be more performant instead.

**Comment by boblail** [comment]
@MikeMcQuaid, would you be receptive to a PR that kept `git cat-file --batch` as a strategy but implemented it in `FormulaVersions` instead of `FormulaeAtRevision`?

**Comment by MikeMcQuaid** [comment]
@boblail Potentially: depends on how much the complexity increases. I'm not really convinced that the solution here is in `git cat-file --batch` usage but more:

**Comment by MikeMcQuaid** [comment]
- checking fewer revisions at all
- checking fewer formulae
- not checking formulae at all in some cases
- caching versions

**Comment by boblail** [comment]
@MikeMcQuaid, what do you think of this refactor?

It
- discards `FormulaeAtRevision`
- extracts `GitRepositoryExtension#git_cat_files`
   (which, I think, could also optimize [formula-auditor](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/formula_auditor.rb#L719-L721) — instead of catting many formulae at one revision, we're catting one formula at many revisions, but I don't want to explode the scope of this PR 😬)
- implements `file_contents_at_revision` in terms of `repository.git_cat_file` and `formula_at_revision` in terms of `FormulaVersions.formula_from_contents` so
    1. there's not a duplication of logic between the one-off and bulk approaches for fetching previous versions of formulae
    2. the new strategy is better covered by other specs that were leveraging `FormulaVersions#formula_at_revision`

File: Library/Homebrew/extend/git_repository.rb [hunk 1]
```diff
@@ -101,6 +102,44 @@ def git_commit_message(commit = "HEAD", safe: false)
...
+    # or else the message "<ENTRY> missing" (which indicates a missing blob)
+    s = StringScanner.new(output)
+    [*entries, nil].each_cons(2).to_h do |(entry, next_entry)|
+      if s.scan(/#{entry} missing\n/)
+        [entry, nil]
+      elsif s.scan(/#{Regexp.escape(snip)}\n/)
+        contents = next_entry ? s.scan_until(/^(?=#{Regexp.escape(snip)}|#{next_entry} missing)/) : s.rest
+        [entry, contents]
+      else
+        raise "that didn't work"
```
Comment (boblail) [inline Library/Homebrew/extend/git_repository.rb:131]: - [ ] Better error message

**Comment by boblail** [comment]
> What's the performance benefit here for the typical and pathological cases?

@MikeMcQuaid, would you consider running `brew update` once a week typical and once every 3 months pathological?

At Square, we nudge engineers to do this monthly and we regularly see outliers who haven't done it in several months.

If so, this branch makes `brew update`
- **3.7×** faster for the typical use-case (22.6s → 6.1s)
- **9.3×** faster for the pathological use-case (135s → 14.5s)

**Comment by MikeMcQuaid** [comment]
@boblail Yeh, that's a decent speedup, thanks. I'm going to want to iterate a lot on the implementation here so I'd rather if we can land something like https://github.com/Homebrew/brew/pull/13243#issuecomment-1118475144 to simplify the need for this instead.

**Comment by MikeMcQuaid** [comment]
I think https://github.com/Homebrew/brew/pull/13299 may remove the need for this but interested in thoughts.

**Comment by MikeMcQuaid** [comment]
Passing on this given recent `brew update` changes. Feel free to reopen if still needed/desired.

**Pull request closed without merging**